src/pages/node-express/lesson13-a1.mdx

```diff
@@ -0,0 +1,526 @@
+3. An object with a `message` property and `type`, this can be used in both error and success cases to show flash messages. Here we'e only using it in non-error failure cases (user not found or wrong credentials used).
+
+In addition to registering our local strategy, we also define `serializeUser` and `deserializeUser` functions on the `passport` object. `serializeUser` is used when saving a user to the request session object. We can't save an object to the session cookie, so we need to "serialize" it a.k.a. encode it as a string. In the above function, we're doing that by telling Passport to save just the user id to the session cookie. Then when Passport is handling a protected route, it will use the `deserializeUser` function to retrieve the user from the database, using the id that was saved to the session cookie.
+
+---
+
+You can now add the following lines to `app.js`, right _after_ the `app.use` for session (Passport
+relies on session):
+
+```js
+const passport = require("passport");
+const passportInit = require("./passport/passportInit");
+
+passportInit();
+app.use(passport.initialize());
+app.use(passport.session());
+```
+
+First we call the `passportInit` function that we created in the previous section. This registers our `local` Passport strategy,
+and sets up the `serializeUser` and `deserializeUser` functions onto the `passport` object. Then we call `passport.initialize()` (which sets up Passport to work with Express and sessions) and `passport.session()` (which sets up an Express middleware that runs on all requests, checks the session cookie for a user id, and if it finds one, deserializes and attaches it to the `req.user` property).
+
+Finally, you can now uncomment the lines having to do with Passport in `routes/sessionRoutes.js`,
+so that the require statement for Passport is included, and so that the route for logon looks like
+
+```js
+router
+  .route("/logon")
+  .get(logonShow)
+  .post(
+    passport.authenticate("local", {
+      successRedirect: "/",
+      failureRedirect: "/sessions/logon",
+      failureFlash: true,
+    })
+  );
+```
+
+This means that when someone sends a `POST` request to the `/sessions/logon` path, Passport will call the function we defined earlier and registered to the name `"local"`. If that function completes successfully (`done(...)` is called with no error argument), then it will redirect the page to the `successRedirect` page. If there is an error, then it will redirect to the `failureRedirect` page, and also set a flash message with the `message` property from the object we passed to `done(...)`.
+
+Since we're letting Passport handle setting the `req.flash` properties now, we can remove the lines in `routes/sessionController.js` that set the flash messages for the `loginShow` handler. So that should now just look like:
+
+```js
+const logonShow = (req, res) => {
+  if (req.user) {
+    return res.redirect("/");
+  }
+  res.render("logon");
+};
+```
```

Comment:
More Passport explaining; and a new code example. ^ Without the above change I was having trouble getting flash messages to work for the logon route

Reply:
I think because once we access `req.flash(<type>)` in the `storeLocals` middleware, then they're no longer available directly on the `req` object.